### PR TITLE
Add ReadChangeStream end time logic, minor cleanup

### PR DESCRIPTION
**Pull request by jackdingilian**
Add end time parameter to finish stream after reading up to the given time. This is not reliable right now so we don't expose it to users yet.

 Additionally update client configuration, add a couple unit tests and do some minor clean up. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by jackdingilian** [comment]
Run Java_GCP_IO_Direct PreCommit

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/action/ReadChangeStreamPartitionAction.java [hunk 1]
```diff
@@ -160,6 +167,22 @@ public ProcessContinuation run(
     // Process CloseStream if it exists
     CloseStream closeStream = tracker.currentRestriction().getCloseStream();
     if (closeStream != null) {
+      // tracker.currentRestriction().closeStream.getStatus()
```
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/action/ReadChangeStreamPartitionAction.java:170]: leftover from development?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java [hunk 2]
```diff
@@ -100,9 +101,30 @@ private static BigtableChangeStreamAccessor createAccessor(@NonNull BigtableConf
     dataSettingsBuilder.setInstanceId(instanceId);
     tableAdminSettingsBuilder.setInstanceId(instanceId);
 
-    if (appProfileId != null) {
-      dataSettingsBuilder.setAppProfileId(appProfileId);
-    }
+    String appProfileId = checkArgumentNotNull(bigtableConfig.getAppProfileId()).get();
```
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java:104]: is `appProfileId` now mandatory?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java [hunk 3]
```diff
@@ -100,9 +101,30 @@ private static BigtableChangeStreamAccessor createAccessor(@NonNull BigtableConf
...
+                        .setMaxChannelCount(100)
+                        // Start off with 5
+                        .setInitialChannelCount(5)
+                        // Make sure the channels are primed before use
+                        .setPreemptiveRefreshEnabled(true)
+                        // evict channels when there are less than 10 outstanding RPCs
+                        .setMinRpcsPerChannel(10)
+                        // add more channels when the channel has 50 outstanding RPCs
+                        .setMaxRpcsPerChannel(50)
+                        .build())
```
Comment (pabloem) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java:126]: just curious - how did you arrive at these numbers? (and the below 10 attempts?) - maybe worth mentioning in comments whether you tested with these and so on

**Comment by pabloem** [comment]
lgtm - just waiting for responses on the comments I left

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java (hunk 2 above)
Comment (jackdingilian) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java:104]: It is currently mandatory, we already had the `checkArgumentNotNull(bigtableConfig.getAppProfileId())` above. We are actually going to update this to default to the default app profile in a follow on though.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java (hunk 3 above)
Comment (jackdingilian) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/dao/BigtableChangeStreamAccessor.java:131]: The channel pool settings are somewhat arbitrary. The main goal is to increase the number of concurrent streams per worker. We know that we expect a max of dataflow `workerHarnessThreads` open streams per worker (bc we use the blocking stream APIs) and that defaults to 500. Before this change, the channel pool size is 2, and there is a max of 100 concurrent streams per channel so we are limiting the number of active streams to 200 effectively.  Added a comment.

The 10 attempts below is to match the default for other bigtable rpcs. We noticed that these were occasionally throwing exceptions for transient failures that should be retried while other methods weren't.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/action/ReadChangeStreamPartitionAction.java (hunk 1 above)
Comment (jackdingilian) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/changestreams/action/ReadChangeStreamPartitionAction.java]: Yes. Removed, thanks

**Comment by jackdingilian** [comment]
Run Java_GCP_IO_Direct PreCommit

**Comment by pabloem** [comment]
Run Java PreCommit

**Comment by JrSchild** [comment]
Hi @jackdingilian and @pabloem,

Much appreciated for your work here. Apologies for digging up this PR. I wondered if you could tell me why the end time parameter is not reliable right now on the BigTable ReadChangeStream, and what needs to happen in order for it to be usable in production? I would really like to leverage this functionality to perform batch processing jobs with dataflow. If this feature is not yet available, do you have any alternatives to achieve similar functionality? (That being, reading the change log stream from and until a given point in time).

Kind regards!

**Comment by jackdingilian** [comment]
Hi @JrSchild.

Unfortunately right now because of the way the estimatedLowWatermark works we are not able to support this use case reliably. If possible, could you please email cloud-bigtable-change-streams-feedback@google.com with some details about the use case you described? I think this will be useful to get more attention for potential future support. Thanks!

**Pull request merged**